specs/vm/instruction_set.md

```diff
@@ -1416,6 +1417,53 @@ Cease VM execution and revert script effects. After a revert:
 1. All [OutputContract](../protocol/tx_format.md#outputcontract) outputs will have the same `balanceRoot` and `stateRoot` as on initialization.
 1. All [OutputVariable](../protocol/tx_format.md#outputvariable) outputs will have `to`, `amount`, and `asset_id` of zero.
 
+### SMO: Send message to output
+
+|             |                                                                                     |
+|-------------|-------------------------------------------------------------------------------------|
+| Description | Send a message to recipient address `MEM[$rA, 32]` with call abi `MEM[$rA + 32, $rB]` and `$rD` coins, with output `$rC`. |
+| Operation   | ```outputmessage(MEM[$fp, 32], MEM[$rA, 32], MEM[$rA + 32, $rB], $rD, $rC);```      |
+| Syntax      | `smo $rA, $rB, $rC, $rD`                                                            |
+| Encoding    | `0x00 rA rB rC rD`                                                                  |
+| Notes       |                                                                                     |
+
+Given helper `balanceOfStart(asset_id: byte[32]) -> uint32` which returns the memory address of the remaining free balance of `asset_id`, or panics if `asset_id` has no free balance remaining.
+
+Panic if:
+
+- `$rA + $rB + 32` overflows
+- `$rA + $rB + 32 > VM_MAX_RAM`
+- `$rB > MEM_MAX_ACCESS_SIZE`
+- `$rB > MESSAGE_MAX_DATA_SIZE`
+- `$rC > tx.outputsCount`
+- In an external context, if `$rD > MEM[balanceOfStart(0), 8]`
+- In an internal context, if `$rD` is greater than the balance of asset ID 0 of output with contract ID `MEM[$fp, 32]`
+- `tx.outputs[$rC].type != OutputType.Message`
+- `tx.outputs[$rC].recipient != 0`
+
+Append a receipt to the list of receipts, modifying `tx.receiptsRoot`:
+
+| name         | type          | description                                                               |
+|--------------|---------------|---------------------------------------------------------------------------|
+| `type`       | `ReceiptType` | `ReceiptType.MessageOut`                                                  |
+| `messageID`  | `byte[32]`    | The messageID as described [here](../protocol/identifiers.md#message-id). |
+| `sender`     | `byte[32]`    | The address of the message sender: `MEM[$fp, 32]`.                        |
+| `recipient`  | `byte[32]`    | The address of the message recipient: `MEM[$rA, 32]`.                     |
+| `amount`     | `uint64`      | Amount of base asset coins sent with message: `$rD`.                      |
+| `nonce`      | `byte[32]`    | The message nonce.                                                        |
```

Comment:
Rather than nonce, it could be simpler to just use the output index, i.e. `$rC`.

Reply:
It needs to be something unique to each message, or else someone malicious could force two messages to have the same `messageID`. I thought we were decided on `hash(txID ++ outputIndex)`. I'll add the concept of `OutputMessageNonce` to the identifiers section and link here

Reply:
Transaction IDs are unique across the entire history of the chain, so by construction `txID ++ outputIndex` is collision-free.

Reply:
The nonce concept is still used when verifying merkle proofs as a way to reduce calldata (provide just a hash vs a byte32 + uint8). I'm leaning towards logging the full nonce rather than the bits to compute the nonce simply out of convenience when fetching the data required to prove a message on L1. Similar to how we log the full messageID even though it could just be computed from everything else.

Reply:
Hmm but the receipt and what's in the Merkle root of outbound messages to the EVM chain are actually independent. They can contain different values. Can you also clarify what specifically this nonce is? Is it just some unique identifier for the message? If so, how does it differ from message ID.

Reply:
The nonce is here to ensure that each messageID is unique. The relay function on the L1 inbox looks like this: `relayMessage(bytes32 sender, bytes32 recipient, bytes32 nonce, uint64 amount, bytes data, BinaryMerkleProof merkleProof)`. The function hashes all values to figure out the `messageID` and then uses the provided merkle proof to verify the `messageID` is valid. If we didn't have the nonce, we'd have to pass in both `byte32 txID` and `uint8 outputIndex` requiring more calldata. The nonce is defined [Here](https://github.com/FuelLabs/fuel-specs/blob/pixelcircuits-refactor-depositwithdraw-to-flat-messaging/specs/protocol/identifiers.md#output-message-nonce)

Reply:
Ah got it. Can you link to that in here? Something like

```
The message [nonce](../protocol/identifiers.md#output-message-nonce).
```

Reply:
Ah this is already done!